File: dijkstra.cpp
#include <vector>
#include <algorithm>

using namespace std;

struct Vertice {
    int nombre;
};

struct Arista {
    Vertice origen;
    Vertice destino;
    int peso;
};

struct Dijkstra {
    vector<Arista> a;
    vector<Vertice> vc;
    vector<Vertice> vnc;
    vector<int> pacu;
    vector<Vertice> prev;
};

const int oo = 100;
Vertice verticenulo = {0};


// Función para calcular el costo entre dos vértices
auto costo = [](const std::vector<Arista>& aristas, const Vertice& origen, const Vertice& destino) {
    auto arista = find_if(aristas.begin(), aristas.end(), [&origen, &destino](const Arista& a) {
        return a.origen.nombre == origen.nombre && a.destino.nombre == destino.nombre;
    });
    //
    return (arista != aristas.end()) ? arista->peso : oo;
};

// Función para inicializar la lista de predecesores
auto previnicial = [](const Vertice& v, const std::vector<Vertice>& vertices) {
    std::vector<Vertice> predecesores(vertices.size(), verticenulo);
    //Aqui se usa transform que es el equivalente a map. Va a aplicar una operacion a cada elemento dentro del rango especificado.
    std::transform(vertices.begin(), vertices.end(), predecesores.begin(), 
                    [&v](const Vertice& x) {
                        return (x.nombre == v.nombre) ? v : verticenulo;
                    });
    return predecesores;
};

// Función para inicializar la lista de pesos acumulados
auto acuinicial = [](const std::vector<Arista>& aristas, const Vertice& v, const std::vector<Vertice>& vertices) {
    std::vector<int> pesosAcumulados(vertices.size(), 0);
    std::transform(vertices.begin(), vertices.end(), pesosAcumulados.begin(),
                    [&aristas, &v](const Vertice& x) {
                        return (x.nombre != v.nombre) ? costo(aristas, v, x) : 0;
                    });
    return pesosAcumulados;
};

// Función para actualizar los pesos acumulados y predecesores
pair<vector<int>, vector<Vertice>> actpeso(const vector<int>& pacu,
                                           const vector<Vertice>& prev,
                                           const Arista& arista) {
    vector<int> nuevosPesosAcumulados = pacu;
    vector<Vertice> nuevosPredecesores = prev;

    int costoNuevo = pacu[arista.origen.nombre - 1] + arista.peso;

    if (costoNuevo < pacu[arista.destino.nombre - 1]) {
        nuevosPesosAcumulados[arista.destino.nombre - 1] = costoNuevo;
        nuevosPredecesores[arista.destino.nombre - 1] = arista.origen;
    }

    return make_pair(nuevosPesosAcumulados, nuevosPredecesores);
}

Dijkstra iteracion(const Dijkstra& d) {
    Dijkstra resultado = d;

    // Encuentra el próximo vértice a considerar
    Vertice next = *(min_element(d.vnc.begin(), d.vnc.end(),
        [&d](const Vertice& x, const Vertice& y) {
            return d.pacu[x.nombre - 1] < d.pacu[y.nombre - 1];
        }));

    // Actualiza la lista de vértices considerados y no considerados
    resultado.vc.insert(resultado.vc.begin(), next);
    resultado.vnc.erase(remove(resultado.vnc.begin(), resultado.vnc.end(), next), resultado.vnc.end());

    // Filtra las aristas candidatas
    vector<Arista> aristasCandidatas;
    for (const Arista& a : resultado.a) {
        if (a.origen.nombre == next.nombre && find(resultado.vnc.begin(), resultado.vnc.end(), a.destino) != resultado.vnc.end()) {
            aristasCandidatas.push_back(a);
        }
    }

    // Actualiza los pesos acumulados y predecesores
    for (const Arista& arista : aristasCandidatas) {
        auto [nuevosPacu, nuevosPrev] = actpeso(resultado.pacu, resultado.prev, arista);
        resultado.pacu = nuevosPacu;
        resultado.prev = nuevosPrev;
    }

    return resultado;
}

File: dijkstra.hs
data Vertice = Vertice {nombre::Int} deriving (Eq)

instance Show (Vertice) where show (Vertice n) = show (n)

data Arista = Arista {orig::Vertice, dest::Vertice, peso::Int} deriving Eq

instance Show (Arista) where show (Arista orig dest peso) = "("++(show orig)++"->"++(show dest)++","++(show peso)++")"

--getOrig (Arista o d ) = o

--getDest (Arista o d) = d

v1 = Vertice   1 
v2 = Vertice   2     
v3 = Vertice   3 
v4 = Vertice   4 
v5 = Vertice   5 

a1 = Arista v1 v2 5
a2 = Arista v1 v3 2
a3 = Arista v1 v4 3
a4 = Arista v2 v5 1
a5 = Arista v3 v2 2 
a6 = Arista v4 v2 1 
a7 = Arista v4 v3 4
a8 = Arista v4 v5 7

verticesprueba = [v1,v2,v3,v4,v5]

aristasprueba = [a1,a2,a3,a4,a5,a6,a7,a8]

data Grafo = Grafo {vertices::[Vertice], aristas::[Arista]} deriving Show

grafoprueba = Grafo verticesprueba aristasprueba

at:: [a] -> Int -> a
at xs i = xs !! (i-1)

type VC = [Vertice]

type VNC = [Vertice]

type Peso = Int

type Pacu = [Peso]

type Prev = [Vertice]

type Aristas = [Arista] 

vercono v = filter (\x->x==v) verticesprueba

vernocono v = filter (\x->x/=v) verticesprueba 

data Dijkstra = Dij {a::Aristas, vc::VC, vnc::VNC, pacu::Pacu, prev::Prev} deriving Show

oo = 100


verticenulo = Vertice 0

costo::Aristas -> Vertice -> Vertice -> Peso
costo as o d = if (filtrado == []) then oo else (peso $ head $ filtrado)
   where filtrado = filter (\a->(orig a == o && dest a == d)) as

--peso $ head $ ( filter (\a->(orig a == o && dest a == d)) as)

previnicial v = map (\x->if (x==v) then v else verticenulo) verticesprueba

acuinicial a v = map (\x->if (x==v) then 0 else (costo a v x)) verticesprueba

cambiarnth:: a -> Int -> [a] -> [a]
cambiarnth a n [] = []
cambiarnth a 0 (x:xs) = a:xs
cambiarnth a n (x:xs) = x:(cambiarnth a (n-1) xs)

generarinicial :: Vertice -> Dijkstra
generarinicial v = Dij (aristasprueba) (vercono v) (vernocono v) (acuinicial (aristasprueba) v) (previnicial v)

getpeso:: Pacu -> Vertice -> Peso
getpeso p v = at p (nombre v) 


actpeso::(Pacu , Prev) -> Aristas ->(Pacu,Prev)
actpeso (p, pr) [] = (p,pr)
actpeso (p, pr) (a:as) = actpeso (nuevocos,nuevopre)   as
   where cambiar  = cambiarnth (costonuevo) (nombre (dest a)-1) p
         cambiarpre = cambiarnth (dest a) (nombre (orig a)-1) pr
         costonuevo = (peso a) + getpeso p (orig a)
         nuevocos  = if ((costonuevo) < getpeso p (dest a)) then cambiar else p 
         nuevopre  = if ((costonuevo) < getpeso p (dest a)) then cambiarpre else pr 


iteracion::Dijkstra -> Dijkstra
iteracion d = Dij (a d) (nuevocon d) (nuevonocono d) (nuevoacu d) (nuevoprev d)
     where next = foldl (\x y->if ((getpeso (pacu d) x) < (getpeso (pacu d) y)) then x else y) (head $ vnc d) (vnc d) 
           nuevocon d = (next):(vc d) 
           nuevonocono d = filter (/=next ) (vnc d)
           acandidata = filter (\x-> orig x == next) (a d)
           avalidas   = filter (\x-> dest x `elem` (vnc d)) acandidata
           resul      = actpeso (pacu d,prev d) avalidas
           nuevoacu d = fst resul 
           nuevoprev d = snd resul  

caminocorto :: Dijkstra -> Dijkstra
caminocorto d  
  | vnc d  == [] = d
  | otherwise = caminocorto $ iteracion d

--caminocorto (Dij a c [] ac pr) = Dij a c [] ac pr
--caminocorto (Dij a c nc ac pr) = 


